js/directoryLinksProvider.js

```diff
@@ -0,0 +1,85 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+ /*globals gDrag, gNewTab, gGrid, gUndoDialog, async, swMessage, gPinnedLinks, gBlockedLinks*/
+
+"use strict";
+
+// The const that tells where to obtain directory links
+const PREF_DIRECTORY_SOURCE = "https://tiles.services.mozilla.com/v3/links/fetch/en-US/nightly";
+
+/**
+ * Emits notifications to PlacesProvider and Links
+ */
+const gDirectoryLinksProvider = {
+  // download default interval is 24 hours in milliseconds
+  _downloadIntervalMS: 86400000,
+
+  // links download deferred, resolved upon download completion
+  _downloadDeferred: null,
+
+  _directoryLinksRequest: new Request(PREF_DIRECTORY_SOURCE),
+
+  init: async(function* () {
+    var self = gDirectoryLinksProvider;
+    gDirectoryLinksProvider._lastDownloadMS = 0;
+    var hasEntry = yield CacheTasks.hasCacheEntry(self._directoryLinksRequest, "directory_links");
+
+    if (hasEntry) {
+      var response = yield CacheTasks.respondFromCache(self._directoryLinksRequest, "directory_links");
+      var txt = yield response.text();
+      self._lastDownloadMS = JSON.parse(txt).timestamp;
```

Comment:
(see my other comment. We should pull this from the HTTP headers of the Response)